### ISSUE TITLE
CPMR0072 is inconsistent with Chocolatey documentation

### ISSUE THREAD

**Issue by Thilas** [chocolatey/docs#541]
### What You Are Seeing?

1. The brand new requirement [CPMR0072](https://github.com/chocolatey/docs/blob/master/input/en-us/community-repository/moderation/package-validator/rules/cpmr0072.md) is inconsistent with the official Chocolatey documentation about [environment variables](https://docs.chocolatey.org/en-us/create/functions/#environment-variables). Indeed, several variables should no longer be used but are still documented as OK to use (e.g. `chocolateyPackageFolder`).
2. This breaking change have been introduced out of the blue, without letting a chance to maintainers to adjust their packages accordingly or discuss the impacts for that matter. As a conclusion, many packages will take way more time to be delivered to the community.
3. IMHO this change will be a big source of troubles for maintainers, at least regarding `chocolateyPackageFolder` and possibly `chocolateyToolsLocation`, both of which have **always** been documented as OK to use BTW.

### What is Expected?

1. You should make sure documentation is consistent with validation rules.
2. You should switch the new requirements that just got published into guidelines in order to give enough time to maintainers to adjust and/or discuss. And then put them back to requirements.
3. I would like you to reconsider this change at least regarding `chocolateyPackageFolder` and `chocolateyToolsLocation` as it seems pretty useful in some cases to be able to know those locations.

### How Did You Get This To Happen? (Steps to Reproduce)

One of my package just got rejected because of the new requirement: https://community.chocolatey.org/packages/vscodium.portable/1.71.0.22245
whereas it was passing like a charm 2 weeks ago: https://community.chocolatey.org/packages/vscodium.portable/1.70.2.22230

Related source code: https://github.com/Thilas/chocolatey-packages/tree/master/vscodium.portable/tools

**Comment by gep13** [comment]
> @Thilas said...
is inconsistent with the official Chocolatey documentation about [environment variables](https://docs.chocolatey.org/en-us/create/functions/#environment-variables).

Thank you for pointing out this inconsistency. This is something that we will definitely want to get updated.

> @Thilas said...
This breaking change have been introduced out of the blue, without letting a chance to maintainers to adjust their packages accordingly or discuss the impacts for that matter. 

We did mention the upcoming changes in a blog post that was announced here: https://blog.chocolatey.org/2022/08/upcoming-changes-validator/ In future, I think that we should provide more time before the change, however, as mentioned in the blog post:

> if there is an impact to a package, the package maintainer can request a temporary exemption to give them time to make changes that will comply with the new rules.

> @Thilas said...
I would like you to reconsider this change at least regarding chocolateyPackageFolder and chocolateyToolsLocation as it seems pretty useful in some cases to be able to know those locations.

Regarding usage of the `chocolateyPackageFolder` environment variable, you can get this information using something like:

```
$toolsDir   = "$(Split-Path -parent $MyInvocation.MyCommand.Definition)"
```

which is what is shown/used within the templated files that you can get from `choco new` command.

For `chocolateyToolsLocation`, the recommended route is using the `Get-ToolsLocation` cmdlet, which it appears as though you are already using.

**Comment by Thilas** [comment]
Thanks @gep13 for the quick answer! At least I know how to quickly fix my issue, though I was about to do exactly that. It's just that it looked to me as a kind of messy workaround.

> We did mention the upcoming changes in a blog post that was announced here: https://blog.chocolatey.org/2022/08/upcoming-changes-validator/ In future, I think that we should provide more time before the change

A blog post is great but many people just won't see it. Instead, having a warning from the automatic validation process for a while (a couple of months I would say, especially during summer) would highly increase the chance of people seing such change coming.

Anyway I'll leave this issue open at least for the documentation to be straighten up if you don't mind.

Thanks again.

**Comment by gep13** [comment]
@Thilas thanks again for raising this!

I have updated the documentation to be consistent with the requirements which are now in place in the package-validator. Let me know if you have any questions. Thanks

**Comment by Thilas** [comment]
All good, my package is now fixed. Thanks.

**Comment by koitsu** [comment]
For what it's worth: I just got bit by this. Everything @Thilas said is correct, all the way down to "A blog post is great but many people won't see it".  In fact, even [packages your own Technical Engineering Manager ( @pauby ) and author of the blog post maintains](https://github.com/pauby/ChocoPackages/blob/master/automatic/keepass-plugin-keetraytotp/tools/chocolateyInstall.ps1) are violating this policy. :)

You also still have documentation that references these variables, such as https://docs.chocolatey.org/en-us/getting-started#where-are-chocolatey-packages-installed-to which talks about and advocates $env:ChocolateyBinRoot.

No offense, but I strongly suggest you do an `egrep -r -i 'chocolateyToolsLocation|chocolateyBinRoot|chocolatey_bin_root|chocolateyPackageFolder|packageFolder|chocolateyChecksum32|chocolateyChecksum64|chocolateyChecksumType32|chocolateyChecksumType64|downloadCacheAvailable'` on your entire documentation repository and fix what comes back.  This (documentation) change should have been done prior to said validator change actually being rolled out.

Additionally, it would be helpful if the CPMR0072 documentation was also more clear in what it's referring to.  When I see the term "private variables" combined with a web page that says things like `chocolateyBinRoot`, I am thinking it's talking about things like `$chocolateyBinRoot` and not `$env:ChocolateyBinRoot` (re: private vs. class-ified).  The "Recommended Solution" section is quite awful and should really give examples/scenarios of situations where each of these variables have been used (historically) and provide recommendations for solving them (ex. "use this helper lib/function instead"), not just "Remove any use of".  For example, this code is now flagged as Bad And Evil even though it's purely informative:

```
if (! $installPath) {
  Write-Verbose "$($packageSearch) not found in $($env:ChocolateyBinRoot)"
  throw "$($packageSearch) install location could not be found."
}
```

**Comment by gep13** [comment]
> @koitsu said...
even [packages your own Technical Engineering Manager ( @pauby ) and author of the blog post maintains](https://github.com/pauby/ChocoPackages/blob/master/automatic/keepass-plugin-keetraytotp/tools/chocolateyInstall.ps1) are violating this policy. :)

I am not sure what bearing this has on anything.  This just means that the next time @pauby pushes a version of that package, that he will need to update it based on the recommendations in this rule.

> @koitsu said...
You also still have documentation that references these variables

This is clearly an oversight, and something which we will get fixed.  If you wanted to help with this, then we also gratefully accept pull requests. All of our documentation is open source in this repository.

> @koitsu said...
The "Recommended Solution" section is quite awful and should really give examples/scenarios of situations where each of these variables have been used (historically) and provide recommendations for solving them (ex. "use this helper lib/function instead"), not just "Remove any use of".

Again, we are more than happy to take suggestions on how our documentation can be improved.  If you wanted to raise a PR with some suggestions, we can definitely get this reviewed and updated in the docs.

> @koitsu said...
For example, this code is now flagged as Bad And Evil even though it's purely informative:

Yes, it is informative, assuming that the environment variable that you are referencing actually contains anything.  This is not guaranteed.  The fact that Chocolatey decides to store the information in an environment variable is an _implementation detail_ and not something that you as a package maintainer can rely on.  That is why we provide the `Get-ToolsLocation` [CmdLet](https://docs.chocolatey.org/en-us/create/functions/get-toolslocation) which we recommend using in these cases.

In addition to the above, I notice that you are using `Get-BinRoot`.  This alias was marked as deprecated in v1.0.0 of Chocolatey CLI, as per the documentation linked above, and also which is output as a warning when it's uses is detected:

https://github.com/chocolatey/choco/blob/develop/src/chocolatey.resources/helpers/functions/Get-ToolsLocation.ps1#L52

And this alias will be removed in v2.0.0 of Chocolatey, so I would recommend that this be changed as well.

**Comment by koitsu** [comment]
We've both stated our opinions on how this matter was handled, as well as how to improve the documentation. I will not comment on those going forward; the conversation I don't think would go anywhere positive, and both of us certainly have better things to exert our energy on than slightly-heated conversations in GitHub.

> Yes, it is informative, assuming that the environment variable that you are referencing actually contains anything. This is not guaranteed. The fact that Chocolatey decides to store the information in an environment variable is an _implementation detail_ and not something that you as a package maintainer can rely on. That is why we provide the `Get-ToolsLocation` [CmdLet](https://docs.chocolatey.org/en-us/create/functions/get-toolslocation) which we recommend using in these cases.
> 
> In addition to the above, I notice that you are using `Get-BinRoot`. This alias was marked as deprecated in v1.0.0 of Chocolatey CLI, as per the documentation linked above, and also which is output as a warning when it's uses is detected:
> 
> https://github.com/chocolatey/choco/blob/develop/src/chocolatey.resources/helpers/functions/Get-ToolsLocation.ps1#L52
> 
> And this alias will be removed in v2.0.0 of Chocolatey, so I would recommend that this be changed as well.

I appreciate this information.  I will have to figure out where this warning message is emitted / during what phase.  I'm thinking during `choco install mypackage -s .` during a local package install/test, and not `choco pack`, but I tend to use `-d -v` during `choco install` so it's very likely its buried deep in a huge amount of output and easily overlooked.  I wouldn't have known this existed had you not pointed it out.

**Comment by pauby** [comment]
I have locked this conversation, but wanted to just pickup on what had been brought up about my packages.

@koitsu There are a couple of my packages that will be hit by the new rule changes. `emacs.portable` failed just yesterday (it was using `$env:ChocolateyPackageFolder`). All of those packages are one that I've picked up off other people to keep them going and not ones I've created. That's not meant as an excuse, just the reason.

I have 153 packages to maintain. I haven't gone through them to see if they will fail and as @gep13 said I'll just fix them when they fail. Chocolatey Team members are not immune to not having perfect packages, so I feel a bit 'seen' here 😄
